### ISSUE TITLE
Add `declaration-block-no-duplicate`

### ISSUE THREAD

**Issue by FloEdelmann** [stylelint/stylelint#6524]
### What is the problem you're trying to solve?

Warn about declaration blocks that can be merged to one:

```scss
.selector-a {
  color: red;
}

// rule should warn about this one;
// it can be merged to the previous one
.selector-b {
  color: red;
}

.selector-c {
  color: green;
}

// rule should NOT warn about this one;
// it could be necessary to have this here for specificity
.selector-d {
  color: red;
}
```

Expected output from autofix:

```scss
.selector-a,
.selector-b {
  color: red;
}

.selector-c {
  color: green;
}

.selector-d {
  color: red;
}
```

### What solution would you like to see?

A new rule `declaration-block-no-duplicate` that warns about consecutive declaration blocks with exactly the same declarations.

**Comment by ybiquitous** [comment]
@FloEdelmann Thanks for the proposal. The new rule idea sounds good to me.

I guess the rule blueprint is below:

- Name: `declaration-block-no-duplicate`
- Primary option: `true`
- Secondary options: ???
- Autofixable: No (we can implement later)
- Message: `Unexpected duplicate declaration block`
- Description: "Disallow duplicate declaration blocks."
- Section: "Avoid errors" -> "Duplicate"

If implementing autofix is not hard, we can include autofix in the first implementation.

Also, I have a question about "consecutive". Considering the consistency with the [`declaration-block-no-duplicate-properties`](https://stylelint.io/user-guide/rules/declaration-block-no-duplicate-properties), it may be clearer to provide the secondary option `ignore: ["consecutive-duplicates"]`.

What do you think?

**Comment by ybiquitous** [comment]
If following our [guidelines](https://stylelint.io/developer-guide/rules#write-the-rule) for adding a rule,

> * make the rule strict by default
> * add secondary ignore options to make the rule more permissive

I believe it might be a better idea that the new rule name would be `declaration-block-no-duplicate` and the secondary option `ignore: ["nonconsecutive-duplicates"]` would be provided.

For example:

### Input

```css
.a { color: red; }
.b { color: red; }
.c { color: blue; }
.d { color: red; }
```

### Output (default)

```css
.a { color: red; }
.b { color: red; } /* ← Unexpected duplicate declaration block */
.c { color: blue; }
.d { color: red; } /* ← Unexpected duplicate declaration block */
```

### Output (`ignore: ["nonconsecutive-duplicates"]`)

```css
.a { color: red; }
.b { color: red; } /* ← Unexpected duplicate declaration block */
.c { color: blue; }
.d { color: red; }
```

Also, `stylelint-config-recommended` would provide the following rule settings:

```js
{
	'declaration-block-no-duplicate': [
		true,
		{
			ignore: ['nonconsecutive-duplicates'],
		},
	],
}
```

like it does for [`declaration-block-no-duplicate-properties`](https://github.com/stylelint/stylelint-config-recommended/blob/b987fbd4da9dff4cdac38d112863563c56879e83/index.js#L12-L17).

What do you think about this suggestion?

**Comment by FloEdelmann** [comment]
Sounds good to me!

**Comment by ybiquitous** [comment]
Again, let me show the rule blueprint below:

- Name: `declaration-block-no-duplicate`
- Primary option: `true`
- Secondary options: `ignore: ["nonconsecutive-duplicates"]`
- Autofixable: No (we can implement later)
- Message: `Unexpected duplicate declaration block`
- Description: "Disallow duplicate declaration blocks."
- Section: "Enforce non-stylistic conventions" -> "Duplicate"

The different things from <https://github.com/stylelint/stylelint/issues/6524#issuecomment-1352375216> are "Secondary options" and "Section". Because I think this rule is for conventions rather than "Avoid errors".

**Comment by FloEdelmann** [comment]
Cool, then the issue label can be changed, right?

**Comment by ybiquitous** [comment]
Let's wait for other members' feedback. Especially, I'd like to hear @jeddy3's opinion.

**Comment by jeddy3** [comment]
>  Because I think this rule is for conventions rather than "Avoid errors".

I think you hit the nail on the head there, @ybiquitous. It feels like a [`*-notation` rule](https://stylelint.io/user-guide/rules#notation) to enforce one representation of something that has multiple:

- single rule with a selector list
- single rule using the `:is()` pseudo-class
- separate rules

```css
.a, b { color: red; }
```

```css
:is(.a, .b) { color: red; }
```

```css
.a { color: red; }
.b { color: red; }
```

Or maybe it should be two rules: one that chooses between separates rules or a selector list, and another that chooses between a selector list or the `:is()` pseudo-class?

**Comment by alexander-akait** [comment]
And what about:
```css
.a { color: red; background: blue; }
.b { color: red; background: green; }
.c { color: blue; }
.d { color: red; }
```

it can be optimized:
```css
.a, .b { color: red }
.a { background: blue; }
.b { background: green; }
.c { color: blue; }
.d { color: red; }
```

Also there are can be problem - browser support:
```css
.foo::before { color: red }
.foo::new-unknown-psedo-element { color: red }
```

In theory we can merge them, but only if your browserslist support `.foo::new-unknown-psedo-element`, otherwise `.foo::before` will be invalid (because CSS invalidate the whole selector list if something is not supported/invalid syntax/etc), to solve it we can use `:is()` (forgiving parsing), but `:is()` is still not forgiving in some browsers, so I think we need more options here

Also there is a problem with sorting:
```css
.a { color: red; background: blue; }
.b { background: green; color: red; }
```

You can sort declaration without any problem and so rule should worh here too.

To be honestly this rule is some strange, it is more about minification and production optimizations, because you should not worry about these things when you write code, because it will be optimized in production, but I agree it can reduce a code and do it more readable in some cases, so I don't have the strong opinion here.

**Comment by ybiquitous** [comment]
@jeddy3 @alexander-akait Thank you so much for the constructive feedback. I missed the `:is()` selector and a [downside of the selector list](https://developer.mozilla.org/en-US/docs/Web/CSS/Selector_list#invalid_selector_list).

The rule to recommend using a potentially invalid selector list may confuse some people. Additionally, applying the rule autofix may cause an error that the whole declaration block becomes invalid.

To mitigate the problem, we can add a new rule to recommend using `:is()`, but people may have to remember to use the 2 rules together.

For example:

### Input

```css
.a { color: red; }
.b { color: red; }
.c, .d { color: red; }
```

### Output (2 rules enabled)

```css
.a { color: red; }
.b { color: red; } /* ← Unexpected duplicate declaration block */
.c, .d { color: red; } /* ← Unexpected duplicate declaration block *//* ← Expected ".c, .d" to be ":is(.c, .d)" */
```

### Output after autofix

```css
.a, .b { color: red; } /* Or, :is(.a, .b) { color: red; } */
:is(.c, .d) { color: red; }
```

(it might depend on the applicable order of rules.)

Or, we may be able to add a secondary option to the `declaration-block-no-duplicate` rule. That option applies `:is(.a, .b)` instead of `.a, .b` for autofix. But I don't feel the option only for autofix is a good design because the rule's responsibility becomes ambiguous (do more than one thing).

At least, I believe we cannot add that rule to our sharable configs (`stylelint-config-recommended` and `stylelint-config-standard`) while there is a concern.

We may need to discuss it more.

**Comment by jeddy3** [comment]
@alexander-akait Thanks for chiming in. All good points, especially the potential to invalidate the whole selector!

Additionally, `is:()` [does not necessarily have equivalent specificity](https://w3c.github.io/csswg-drafts/selectors/#matches-pseudo).  We have one rule ([`selector-not-notation`](https://stylelint.io/user-guide/rules/selector-not-notation)) that can change specificity, but a rule that combines CSS rules could have more impact on the author.

With the considerations we've highlighted above, I wonder whether [a plugin](https://stylelint.io/developer-guide/plugins) is better as I don't think a built-in rule will be [generally useful](https://stylelint.io/user-guide/rules). For example, `normalize.css` [chooses not to combine declaration blocks](https://github.com/csstools/normalize.css/blob/main/normalize.css#L110-L159). I suspect most users will use whichever representation is most appropriate on a case-by-case basis; whether that's separate rules, selector lists, or forgiving selector lists within the `:is()` or `:where()` pseudo-classes. 

If they prefer, authors can choose to craft a plugin to universally enforce a particular representation based on the their specific criteria while taking into account the potential to invalidate selectors and change specificity in their codebase. It'll be easier for them to craft something specific to their needs than for us to try to create something that's general-purpose.

**Comment by ybiquitous** [comment]
I agree that the new rule is only useful for specific use cases and unsuitable for built-in rules.

Or perhaps it would be better to say that it is difficult to create such a rule that would satisfy a generic use case for duplicate declaration blocks.

**Comment by jeddy3** [comment]
Thank you everyone for working this through. Closing as better suited as a plugin.

@FloEdelmann Thank you for creating the issue. You can [craft a plugin](https://stylelint.io/developer-guide/plugins/) to meet your needs.
